### DISCUSSION TITLE
PicoGK and Simulation

### DISCUSSION THREAD

**Discussion by josefinelissner** [leap71/PicoGK#32] [General]
Hi, everyone!

I am opening up this discussion, since we are receiving many questions about how PicoGK can be integrated with simulation tools. With PicoGK v1.6 we enabled access to VDB metadata, vector and scalar fields (in addition to just voxelfields). We see a very elegant, possible solution in using one single VDB file to encode all data that is needed for a simulation. I have published a first attempt here: 

https://github.com/leap71/PicoGK_SimulationExample

It is an example project showing how all necessary information for a simple fluid flow simulation can be written and read from one VDB file, that includes fluid and solid domain as voxelfields, the fluid density and viscosity distribution as scalar fields and the initial state of the fluid velocity (this includes inlet boundary patch as you would do it traditionally) as a vector field.

At this stage we are looking for people with knowledge in the field of simulation, ideally you are developing your own tool, to give us constructive feedback. We are aware of the fact that this is not following the traditional approach with volumetric meshing and boundary surface patches. A voxel-based approach obviously tailors better to simulation methods that run on regular grids like the Lattice-Boltzmann methods. However, from a Computational Engineering perspective, the geometry of an object and its physical behavior should not be two completely separate things. CEMs have a lot of metadata about physical constraints anyway, so driving an automated simulation setup would be rather natural, without the need for someone to manually click and re-assign boundary conditions...

Please check it out!

![image](https://github.com/leap71/PicoGK/assets/44703282/c0446b05-fc4b-40fd-aad8-cf7604d1bf5e)

As a teaser: A mesh deformation preview of one of our rover wheels.

<img width="812" alt="Screenshot 2024-05-24 at 9 00 31 PM" src="https://github.com/leap71/PicoGK/assets/44703282/d253b02e-356f-410f-bdd3-ea97bf77fc29">

**Comment by jigarmech** [comment]
Hi Josefine, thanks for the simulation feature. This is what I was looking for and will be trying with my work here to have LBM coupled with the geometry. Next, I will be trying to find those voxels that has velocity magnitude less than certain amount which I would like to remove it from the geometry. I have few other things in my mind as well so will be working on this more.
 
I downloaded it and was trying to apply the code, I believe the `ScalarField `and `VectorField` class are missing. I cannot find them in the code. I maybe wrong but can you look into it as I checked and cannot find them. Am I missing something here? 
Thanks a lot again!

**Comment by LinKayser** [comment]
Hi Jigar, these classes are part of PicoGK, and are included as submodule in the Github repo. If you downloaded it via the ZIP, you have to copy PicoGK from your PicoGK installation. If you are cloning the repo, make sure you init all the submodules.

Lin

**Reply by jigarmech** [reply]
Thanks for your quick response. It seems those files were missing in my PicoGK folder. Updated it. Thanks Lin!

**Reply by LinKayser** [reply]
Awesome!
